src/src/com/microsoft/aad/adal/StorageHelper.java

```diff
@@ -305,71 +280,204 @@ public String decrypt(String value) throws GeneralSecurityException, IOException
-     * @param clearText
+     * Supported API >= 18 PrivateKey is stored in AndroidKeyStore. Loads key
+     * from the file if it exists. If not exist, it will generate one.
+     * 
      * @return
+     * @throws IOException
      * @throws GeneralSecurityException
-     * @throws UnsupportedEncodingException
      */
-    public String encrypt(String clearText)
-            throws GeneralSecurityException, IOException {
+    @TargetApi(Build.VERSION_CODES.JELLY_BEAN_MR2)
+    private final synchronized SecretKey getSecretKeyFromAndroidKeyStore() throws IOException,
+            GeneralSecurityException {
 
-        Logger.v(TAG, "Starting encryption");
+        // Loading file and unwrapping this key is causing performance issue.
+        if (mSecretKeyFromAndroidKeyStore != null) {
+            return mSecretKeyFromAndroidKeyStore;
+        }
 
-        if (StringExtensions.IsNullOrBlank(clearText)) {
-            throw new IllegalArgumentException("Input is empty or null");
+        if (mKeyPair == null) {
+            mKeyPair = getKeyPairFromAndroidKeyStoreForEncryption();
+            Logger.v(TAG, "Retrived keypair from androidKeyStore");
         }
 
-        // load key for encryption if not loaded
-        loadSecretKeyForAPI();
-        Logger.v(TAG, "Encrypt version:" + mBlobVersion);
-        final byte[] blobVersion = mBlobVersion.getBytes(AuthenticationConstants.ENCODING_UTF8);
-        final byte[] bytes = clearText.getBytes(AuthenticationConstants.ENCODING_UTF8);
+        // Store secret key in a file after wrapping
+        File keyFile = new File(mContext.getDir(mContext.getPackageName(), Context.MODE_PRIVATE),
+                ADALKS);
+        Cipher wrapCipher = Cipher.getInstance(WRAP_ALGORITHM);
+        // If keyfile does not exist, it needs to generate one
+        if (!keyFile.exists()) {
+            Logger.v(TAG, "Key file does not exists");
+            final SecretKey key = generateSecretKey();
+            Logger.v(TAG, "Wrapping SecretKey");
+            final byte[] keyWrapped = wrap(wrapCipher, key);
+            Logger.v(TAG, "Writing SecretKey");
+            writeKeyData(keyFile, keyWrapped);
+            Logger.v(TAG, "Finished writing SecretKey");
+        }
+        
+        mSecretKeyFromAndroidKeyStore = getUnwrappedSecretKey(keyFile);
+        return mSecretKeyFromAndroidKeyStore;
+    }
 
-        // IV: Initialization vector that is needed to start CBC
-        byte[] iv = new byte[DATA_KEY_LENGTH];
-        mRandom.nextBytes(iv);
-        IvParameterSpec ivSpec = new IvParameterSpec(iv);
+    /**
+     * Get key pair from AndroidKeyStore.
+     * 
+     * @return
+     * @throws GeneralSecurityException
+     * @throws IOException
+     */
+    @TargetApi(Build.VERSION_CODES.JELLY_BEAN_MR2)
+    private synchronized KeyPair getKeyPairFromAndroidKeyStoreForEncryption() throws GeneralSecurityException, IOException {
+        final KeyStore keyStore = KeyStore.getInstance("AndroidKeyStore");
```

Comment:
same comment here. It would be better to call `readSecretKeyFromKeyStoreForDecryption` if it fails fallback to generate new key